File: app/src/main/java/com/uccd3223/p1_chai_boon_hong_2206806/util/ExerciseGeneratorUtil.java
package com.uccd3223.p1_chai_boon_hong_2206806.util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ThreadLocalRandom;

public final class ExerciseGeneratorUtil {
    public enum SortDirection { ASCENDING, DESCENDING }

    public enum RecognitionPromptMode { WORD_TO_NUMBER, NUMBER_TO_WORD }

    public static final class PlaceValueData {
        private final int tens;
        private final int ones;

        public PlaceValueData(int tens, int ones) {
            this.tens = tens;
            this.ones = ones;
        }

        public int getTens() {
            return tens;
        }

        public int getOnes() {
            return ones;
        }

        public int getTotal() {
            return tens * 10 + ones;
        }

        @Override
        public boolean equals(Object other) {
            if (this == other) return true;
            if (!(other instanceof PlaceValueData)) return false;
            PlaceValueData data = (PlaceValueData) other;
            return tens == data.tens && ones == data.ones;
        }

        @Override
        public int hashCode() {
            return Objects.hash(tens, ones);
        }
    }

    public static final class RecognitionData {
        private final int targetNumber;
        private final List<Integer> options;

        public RecognitionData(int targetNumber, List<Integer> options) {
            this.targetNumber = targetNumber;
            this.options = new ArrayList<>(options);
        }

        public int getTargetNumber() {
            return targetNumber;
        }

        public List<Integer> getOptions() {
            return new ArrayList<>(options);
        }

        @Override
        public boolean equals(Object other) {
            if (this == other) return true;
            if (!(other instanceof RecognitionData)) return false;
            RecognitionData data = (RecognitionData) other;
            return targetNumber == data.targetNumber && options.equals(data.options);
        }

        @Override
        public int hashCode() {
            return Objects.hash(targetNumber, options);
        }
    }

    public static final class RecognitionExercise {
        private final int targetNumber;
        private final List<Integer> options;
        private final RecognitionPromptMode promptMode;

        public RecognitionExercise(
                int targetNumber,
                List<Integer> options,
                RecognitionPromptMode promptMode
        ) {
            this.targetNumber = targetNumber;
            this.options = new ArrayList<>(options);
            this.promptMode = Objects.requireNonNull(promptMode);
        }

        public int getTargetNumber() {
            return targetNumber;
        }

        public List<Integer> getOptions() {
            return new ArrayList<>(options);
        }

        public RecognitionPromptMode getPromptMode() {
            return promptMode;
        }

        @Override
        public boolean equals(Object other) {
            if (this == other) return true;
            if (!(other instanceof RecognitionExercise)) return false;
            RecognitionExercise exercise = (RecognitionExercise) other;
            return targetNumber == exercise.targetNumber
                    && options.equals(exercise.options)
                    && promptMode == exercise.promptMode;
        }

        @Override
        public int hashCode() {
            return Objects.hash(targetNumber, options, promptMode);
        }
    }

    public static final class SortExercise {
        private final List<Integer> options;
        private final SortDirection direction;

        public SortExercise(List<Integer> options, SortDirection direction) {
            this.options = new ArrayList<>(options);
            this.direction = Objects.requireNonNull(direction);
        }

        public List<Integer> getOptions() {
            return new ArrayList<>(options);
        }

        public SortDirection getDirection() {
            return direction;
        }

        public List<Integer> getOrderedNumbers() {
            List<Integer> ordered = new ArrayList<>(options);
            ordered.sort(direction == SortDirection.ASCENDING
                    ? Integer::compareTo
                    : Collections.reverseOrder());
            return ordered;
        }

        @Override
        public boolean equals(Object other) {
            if (this == other) return true;
            if (!(other instanceof SortExercise)) return false;
            SortExercise exercise = (SortExercise) other;
            return options.equals(exercise.options) && direction == exercise.direction;
        }

        @Override
        public int hashCode() {
            return Objects.hash(options, direction);
        }
    }

    public static final class SequenceData {
        private final int[] sequence;
        private final int missingIndex;

        public SequenceData(int[] sequence, int missingIndex) {
            this.sequence = sequence.clone();
            this.missingIndex = missingIndex;
        }

        public int[] getSequence() {
            return sequence.clone();
        }

        public int getMissingIndex() {
            return missingIndex;
        }

        public int getMissingValue() {
            return sequence[missingIndex];
        }

        @Override
        public boolean equals(Object other) {
            if (this == other) return true;
            if (!(other instanceof SequenceData)) return false;
            SequenceData data = (SequenceData) other;
            return missingIndex == data.missingIndex && Arrays.equals(sequence, data.sequence);
        }

        @Override
        public int hashCode() {
            return 31 * Arrays.hashCode(sequence) + missingIndex;
        }
    }

    private ExerciseGeneratorUtil() {
    }

    public static int generateObjectCount(int maxObjects) {
        if (maxObjects < 1) throw new IllegalArgumentException("maxObjects must be at least 1");
        return ThreadLocalRandom.current().nextInt(1, maxObjects + 1);
    }

    public static PlaceValueData generatePlaceValue(int maxTens) {
        if (maxTens < 1) throw new IllegalArgumentException("maxTens must be at least 1");
        return new PlaceValueData(
                ThreadLocalRandom.current().nextInt(1, maxTens + 1),
                ThreadLocalRandom.current().nextInt(0, 10)
        );
    }

    public static RecognitionData generateRecognitionOptions(
            int targetNumber,
            int maxRange,
            int numOptions
    ) {
        if (maxRange < 0) throw new IllegalArgumentException("maxRange must not be negative");
        if (targetNumber < 0 || targetNumber > maxRange) {
            throw new IllegalArgumentException("targetNumber must be inside maxRange");
        }
        if (numOptions < 1 || numOptions > maxRange + 1) {
            throw new IllegalArgumentException("numOptions must fit inside the learning range");
        }

        Set<Integer> preferred = new LinkedHashSet<>();
        if (targetNumber >= 10) preferred.add((targetNumber % 10) * 10 + targetNumber / 10);
        preferred.add(targetNumber - 10);
        preferred.add(targetNumber + 10);
        preferred.add(targetNumber - 2);
        preferred.add(targetNumber + 2);
        preferred.add(targetNumber - 1);
        preferred.add(targetNumber + 1);

        Set<Integer> options = new LinkedHashSet<>();
        options.add(targetNumber);
        List<Integer> preferredList = new ArrayList<>(preferred);
        Collections.shuffle(preferredList);
        for (int candidate : preferredList) {
            if (candidate != targetNumber && candidate >= 0 && candidate <= maxRange) {
                if (options.size() < numOptions) options.add(candidate);
            }
        }

        List<Integer> remaining = new ArrayList<>();
        for (int value = 0; value <= maxRange; value++) {
            if (!options.contains(value)) remaining.add(value);
        }
        Collections.shuffle(remaining);
        for (int candidate : remaining) {
            if (options.size() < numOptions) options.add(candidate);
        }

        List<Integer> shuffled = new ArrayList<>(options);
        Collections.shuffle(shuffled);
        return new RecognitionData(targetNumber, shuffled);
    }

    public static RecognitionExercise generateRecognitionExercise() {
        int target = ThreadLocalRandom.current().nextInt(0, 100);
        RecognitionPromptMode mode = ThreadLocalRandom.current().nextBoolean()
                ? RecognitionPromptMode.WORD_TO_NUMBER
                : RecognitionPromptMode.NUMBER_TO_WORD;
        return generateRecognitionExercise(99, 4, target, mode);
    }

    public static RecognitionExercise generateRecognitionExercise(
            int maxRange,
            int numOptions,
            int targetNumber,
            RecognitionPromptMode promptMode
    ) {
        RecognitionData data = generateRecognitionOptions(targetNumber, maxRange, numOptions);
        return new RecognitionExercise(data.getTargetNumber(), data.getOptions(), promptMode);
    }

    public static SequenceData generateSequence(int start, int step, int length) {
        if (length < 1) throw new IllegalArgumentException("length must be at least 1");
        int[] sequence = new int[length];
        for (int index = 0; index < length; index++) sequence[index] = start + index * step;
        return new SequenceData(sequence, ThreadLocalRandom.current().nextInt(length));
    }

    public static String numberToWords(int number) {
        if (number == 0) return "Zero";
        if (number < 0 || number > 99) return Integer.toString(number);
        String[] ones = {
                "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
                "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen",
                "Seventeen", "Eighteen", "Nineteen"
        };
        String[] tens = {
                "", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"
        };
        if (number < 20) return ones[number];
        String tensWord = tens[number / 10];
        String onesWord = ones[number % 10];
        return onesWord.isEmpty() ? tensWord : tensWord + "-" + onesWord;
    }

    public static List<Integer> generateSortSequence(int length) {
        return generateSortSequence(length, 99);
    }

    public static List<Integer> generateSortSequence(int length, int maxNumber) {
        if (maxNumber < 1) throw new IllegalArgumentException("maxNumber must be at least 1");
        if (length < 1 || length > maxNumber) {
            throw new IllegalArgumentException("length must be between 1 and maxNumber");
        }
        Set<Integer> numbers = new LinkedHashSet<>();
        while (numbers.size() < length) {
            numbers.add(ThreadLocalRandom.current().nextInt(1, maxNumber + 1));
        }
        return new ArrayList<>(numbers);
    }

    public static SortExercise generateSortExercise(int length) {
        SortDirection direction = ThreadLocalRandom.current().nextBoolean()
                ? SortDirection.ASCENDING
                : SortDirection.DESCENDING;
        return generateSortExercise(length, 0, 99, direction);
    }

    public static SortExercise generateSortExercise(
            int length,
            int minNumber,
            int maxNumber,
            SortDirection direction
    ) {
        if (minNumber < 0) throw new IllegalArgumentException("minNumber must not be negative");
        if (maxNumber > 99) throw new IllegalArgumentException("maxNumber must be no more than 99");
        if (minNumber > maxNumber) {
            throw new IllegalArgumentException("minNumber must not exceed maxNumber");
        }
        if (length < 1 || length > maxNumber - minNumber + 1) {
            throw new IllegalArgumentException("length must fit inside the learning range");
        }

        Set<Integer> numbers = new LinkedHashSet<>();
        while (numbers.size() < length) {
            numbers.add(ThreadLocalRandom.current().nextInt(minNumber, maxNumber + 1));
        }
        List<Integer> options = new ArrayList<>(numbers);
        Collections.shuffle(options);
        return new SortExercise(options, direction);
    }

    public static List<Integer> balloonRowPattern(int count) {
        switch (count) {
            case 3:
                return Collections.singletonList(3);
            case 4:
                return Arrays.asList(2, 2);
            case 5:
                return Arrays.asList(3, 2);
            case 6:
                return Arrays.asList(3, 3);
            default:
                throw new IllegalArgumentException("Balloon count must be between 3 and 6");
        }
    }
}

File: app/src/main/java/com/uccd3223/p1_chai_boon_hong_2206806/util/ExerciseGeneratorUtil.kt
package com.uccd3223.p1_chai_boon_hong_2206806.util

import kotlin.random.Random

object ExerciseGeneratorUtil {

    enum class SortDirection { ASCENDING, DESCENDING }

    enum class RecognitionPromptMode { WORD_TO_NUMBER, NUMBER_TO_WORD }

    data class PlaceValueData(val tens: Int, val ones: Int) {
        val total: Int get() = (tens * 10) + ones
    }

    data class RecognitionData(val targetNumber: Int, val options: List<Int>)

    data class RecognitionExercise(
        val targetNumber: Int,
        val options: List<Int>,
        val promptMode: RecognitionPromptMode
    )

    data class SortExercise(
        val options: List<Int>,
        val direction: SortDirection
    ) {
        val orderedNumbers: List<Int>
            get() = when (direction) {
                SortDirection.ASCENDING -> options.sorted()
                SortDirection.DESCENDING -> options.sortedDescending()
            }
    }

    data class SequenceData(val sequence: IntArray, val missingIndex: Int) {
        val missingValue: Int get() = sequence[missingIndex]

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is SequenceData) return false
            return sequence.contentEquals(other.sequence) && missingIndex == other.missingIndex
        }

        override fun hashCode(): Int {
            var result = sequence.contentHashCode()
            result = 31 * result + missingIndex
            return result
        }
    }

    /**
     * 1. Number to Objects Association
     * Generates a random number of objects to display.
     */
    fun generateObjectCount(maxObjects: Int): Int {
        require(maxObjects >= 1) { "maxObjects must be at least 1" }
        return Random.nextInt(1, maxObjects + 1)
    }

    /**
     * 2. Place Value
     * Generates a two-digit number broken down into tens and ones.
     */
    fun generatePlaceValue(maxTens: Int): PlaceValueData {
        require(maxTens >= 1) { "maxTens must be at least 1" }
        val tens = Random.nextInt(1, maxTens + 1)
        val ones = Random.nextInt(0, 10)
        return PlaceValueData(tens, ones)
    }

    /**
     * 3. Numbers Recognition
     * Generates a target number and a shuffled list of distractor options that are similar (close) to the target.
     */
    fun generateRecognitionOptions(targetNumber: Int, maxRange: Int, numOptions: Int): RecognitionData {
        require(maxRange >= 0) { "maxRange must not be negative" }
        require(targetNumber in 0..maxRange) { "targetNumber must be inside maxRange" }
        require(numOptions in 1..(maxRange + 1)) {
            "numOptions must fit inside the learning range"
        }

        val preferred = linkedSetOf<Int>()
        if (targetNumber >= 10) {
            preferred += (targetNumber % 10) * 10 + (targetNumber / 10)
        }
        preferred += targetNumber - 10
        preferred += targetNumber + 10
        preferred += targetNumber - 2
        preferred += targetNumber + 2
        preferred += targetNumber - 1
        preferred += targetNumber + 1

        val options = linkedSetOf(targetNumber)
        preferred
            .filter { it != targetNumber && it in 0..maxRange }
            .shuffled()
            .forEach { if (options.size < numOptions) options += it }

        (0..maxRange)
            .filter { it !in options }
            .shuffled()
            .forEach { if (options.size < numOptions) options += it }

        return RecognitionData(targetNumber, options.shuffled())
    }

    fun generateRecognitionExercise(
        maxRange: Int = 99,
        numOptions: Int = 4,
        targetNumber: Int = Random.nextInt(0, maxRange + 1),
        promptMode: RecognitionPromptMode = if (Random.nextBoolean()) {
            RecognitionPromptMode.WORD_TO_NUMBER
        } else {
            RecognitionPromptMode.NUMBER_TO_WORD
        }
    ): RecognitionExercise {
        val data = generateRecognitionOptions(targetNumber, maxRange, numOptions)
        return RecognitionExercise(data.targetNumber, data.options, promptMode)
    }

    /**
     * 4. Numbers in Sequence
     * Generates an array representing a sequence and the index of the missing element.
     */
    fun generateSequence(start: Int, step: Int, length: Int): SequenceData {
        require(length >= 1) { "length must be at least 1" }
        val sequence = IntArray(length) { i -> start + (i * step) }
        val missingIndex = Random.nextInt(length)
        return SequenceData(sequence, missingIndex)
    }

    fun numberToWords(number: Int): String {
        if (number == 0) return "Zero"
        if (number < 0 || number > 99) return number.toString()
        val ones = arrayOf("", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
            "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen")
        val tens = arrayOf("", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety")

        return if (number < 20) {
            ones[number]
        } else {
            val t = tens[number / 10]
            val o = ones[number % 10]
            if (o.isEmpty()) t else "$t-$o"
        }
    }

    fun generateSortSequence(length: Int, maxNumber: Int = 99): List<Int> {
        require(maxNumber >= 1) { "maxNumber must be at least 1" }
        require(length in 1..maxNumber) { "length must be between 1 and maxNumber" }
        val numbers = mutableSetOf<Int>()
        while (numbers.size < length) {
            numbers.add(Random.nextInt(1, maxNumber + 1))
        }
        return numbers.toList()
    }

    fun generateSortExercise(
        length: Int,
        minNumber: Int = 0,
        maxNumber: Int = 99,
        direction: SortDirection = if (Random.nextBoolean()) {
            SortDirection.ASCENDING
        } else {
            SortDirection.DESCENDING
        }
    ): SortExercise {
        require(minNumber >= 0) { "minNumber must not be negative" }
        require(maxNumber <= 99) { "maxNumber must be no more than 99" }
        require(minNumber <= maxNumber) { "minNumber must not exceed maxNumber" }
        require(length in 1..(maxNumber - minNumber + 1)) {
            "length must fit inside the learning range"
        }

        val numbers = mutableSetOf<Int>()
        while (numbers.size < length) {
            numbers += Random.nextInt(minNumber, maxNumber + 1)
        }
        return SortExercise(numbers.toList().shuffled(), direction)
    }

    fun balloonRowPattern(count: Int): List<Int> = when (count) {
        3 -> listOf(3)
        4 -> listOf(2, 2)
        5 -> listOf(3, 2)
        6 -> listOf(3, 3)
        else -> throw IllegalArgumentException("Balloon count must be between 3 and 6")
    }
}
